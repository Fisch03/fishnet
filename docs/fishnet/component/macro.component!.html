<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.component.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.component.html">macro.component.html</a>...</p>
    <script>location.replace("macro.component.html" + location.search + location.hash);</script>
</body>
</html>