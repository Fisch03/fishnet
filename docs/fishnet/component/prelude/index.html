<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="commonly used imports for building components."><title>fishnet::component::prelude - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../static.files/rustdoc-2c208a72533b4dd0.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="fishnet" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (ccfcd950b 2024-04-15)" data-channel="nightly" data-search-js="search-ffac13a0df2b1870.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../static.files/storage-e32f0c247825364d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-c97aec732c613ca4.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../fishnet/index.html">fishnet</a><span class="version">0.1.0</span></h2></div><h2 class="location"><a href="#">Module prelude</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Aliases</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section><h2><a href="../index.html">In fishnet::component</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../../fishnet/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../index.html">fishnet</a>::<wbr><a href="../index.html">component</a>::<wbr><a class="mod" href="#">prelude</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../../src/fishnet/component/prelude.rs.html#1-26">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>commonly used imports for building components.</p>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.BuildableComponent"><code>pub use super::<a class="trait" href="../trait.BuildableComponent.html" title="trait fishnet::component::BuildableComponent">BuildableComponent</a>;</code></div></li><li><div class="item-name" id="reexport.ComponentState"><code>pub use super::<a class="struct" href="../struct.ComponentState.html" title="struct fishnet::component::ComponentState">ComponentState</a>;</code></div></li><li><div class="item-name" id="reexport.component"><code>pub use crate::<a class="mod" href="../index.html" title="mod fishnet::component">component</a>;</code></div></li><li><div class="item-name" id="reexport.ScriptType"><code>pub use crate::js::<a class="enum" href="../../js/enum.ScriptType.html" title="enum fishnet::js::ScriptType">ScriptType</a>;</code></div></li><li><div class="item-name" id="reexport.css"><code>pub use crate::<a class="mod" href="../../css/index.html" title="mod fishnet::css">css</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="routing/index.html" title="mod fishnet::component::prelude::routing">routing</a></div><div class="desc docblock-short">Routing between [<code>Service</code>]s and handlers.</div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.c.html" title="macro fishnet::component::prelude::c">c</a></div><div class="desc docblock-short">add components to the page.</div></li><li><div class="item-name"><a class="macro" href="macro.css.html" title="macro fishnet::component::prelude::css">css</a></div><div class="desc docblock-short">macro for generating a <a href="../../css/struct.StyleFragment.html" title="struct fishnet::css::StyleFragment"><code>StyleFragment</code></a> from css.</div></li><li><div class="item-name"><a class="macro" href="macro.html.html" title="macro fishnet::component::prelude::html">html</a></div><div class="desc docblock-short">macro for generating <a href="../../type.Markup.html" title="type fishnet::Markup"><code>Markup</code></a> from html.</div></li><li><div class="item-name"><a class="macro" href="macro.script.html" title="macro fishnet::component::prelude::script">script</a></div><div class="desc docblock-short">add inline js to the page</div></li><li><div class="item-name"><a class="macro" href="macro.state.html" title="macro fishnet::component::prelude::state">state</a></div></li><li><div class="item-name"><a class="macro" href="macro.state_init.html" title="macro fishnet::component::prelude::state_init">state_init</a></div></li><li><div class="item-name"><a class="macro" href="macro.style.html" title="macro fishnet::component::prelude::style">style</a></div><div class="desc docblock-short">add <a href="../../macro.css.html" title="macro fishnet::css"><code>css</code></a> to the page</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Arc.html" title="struct fishnet::component::prelude::Arc">Arc</a></div><div class="desc docblock-short">A thread-safe reference-counting pointer. ‘Arc’ stands for ‘Atomically
Reference Counted’.</div></li><li><div class="item-name"><a class="struct" href="struct.Extension.html" title="struct fishnet::component::prelude::Extension">Extension</a></div><div class="desc docblock-short">Extractor and response for extensions.</div></li><li><div class="item-name"><a class="struct" href="struct.Mutex.html" title="struct fishnet::component::prelude::Mutex">Mutex</a></div><div class="desc docblock-short">An asynchronous <code>Mutex</code>-like type.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.FutureExt.html" title="trait fishnet::component::prelude::FutureExt">FutureExt</a></div><div class="desc docblock-short">An extension trait for <code>Future</code>s that provides a variety of convenient
adapters.</div></li><li><div class="item-name"><a class="trait" href="trait.Render.html" title="trait fishnet::component::prelude::Render">Render</a></div><div class="desc docblock-short">Represents a type that can be rendered as HTML.</div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.BoxFuture.html" title="type fishnet::component::prelude::BoxFuture">BoxFuture</a></div><div class="desc docblock-short">An owned dynamically typed <a href="https://doc.rust-lang.org/nightly/core/future/future/trait.Future.html" title="trait core::future::future::Future"><code>Future</code></a> for use in cases where you can’t
statically type your result or need to add some indirection.</div></li><li><div class="item-name"><a class="type" href="type.Markup.html" title="type fishnet::component::prelude::Markup">Markup</a></div><div class="desc docblock-short">A block of markup is a string that does not need to be escaped.</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.component.html" title="attr fishnet::component::prelude::component">component</a></div><div class="desc docblock-short">attribute macro for creating new components</div></li><li><div class="item-name"><a class="attr" href="attr.dyn_component.html" title="attr fishnet::component::prelude::dyn_component">dyn_component</a></div><div class="desc docblock-short">same as <a href="../../attr.component.html" title="attr fishnet::component"><code>component</code></a>, but forces the component to be rerendered each page visit.</div></li></ul></section></div></main></body></html>