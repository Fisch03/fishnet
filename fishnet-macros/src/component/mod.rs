mod parse;
pub(crate) use parse::{parse, parse_dyn};
