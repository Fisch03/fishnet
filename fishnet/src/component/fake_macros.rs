// these are macros that are used by the component macro. when used outside, these fake_macros will
// expand to compile_errors
use fishnet_macros::fake_macro;
fake_macro!(state);
fake_macro!(state_init);
