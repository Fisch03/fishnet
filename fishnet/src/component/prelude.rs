//! commonly used imports for building components.

pub use crate::c;

// components itself
pub use super::{BuildableComponent, ComponentState};
pub use crate::component::component;

// html, js, css
pub use crate::js::ScriptType;
pub use fishnet_macros::css;
pub use maud::{html, Markup, Render};

// boxing runner futures
pub use futures::future::{BoxFuture, FutureExt};

// sharing state
pub use std::sync::Arc;
pub use tokio::sync::Mutex;
// extracting state
pub use axum::Extension;

// endpoints
pub use axum::routing;
