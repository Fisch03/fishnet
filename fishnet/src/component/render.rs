use super::ComponentState;
use async_trait::async_trait;
use futures::future::BoxFuture;
use maud::Markup;

pub type ContentRenderer<ST> =
    Box<dyn Fn(ComponentState<ST>) -> BoxFuture<'static, Markup> + Send + Sync>;

pub struct StatefulContentRenderer<ST>
where
    ST: Clone + Send + Sync,
{
    renderer: ContentRenderer<ST>,
    state: ComponentState<ST>,
}
impl<ST> StatefulContentRenderer<ST>
where
    ST: Clone + Send + Sync,
{
    pub fn new(renderer: ContentRenderer<ST>, state: ComponentState<ST>) -> Box<Self> {
        Box::new(Self { renderer, state })
    }
}

#[async_trait]
pub trait StatefulRenderer: Send + Sync {
    async fn render(&self) -> Markup;
}

#[async_trait]
impl<ST> StatefulRenderer for StatefulContentRenderer<ST>
where
    ST: Clone + Send + Sync,
{
    async fn render(&self) -> Markup {
        (self.renderer)(self.state.clone()).await
    }
}

pub enum ContentType {
    Dynamic(Box<dyn StatefulRenderer>),
    Static(Markup),
}
impl ContentType {
    pub async fn render(&self) -> Markup {
        match self {
            ContentType::Dynamic(renderer) => renderer.render().await,
            ContentType::Static(content) => content.clone(),
        }
    }
}
impl std::fmt::Debug for ContentType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ContentType::Dynamic(_) => write!(f, "Dynamic"),
            ContentType::Static(_) => write!(f, "Static"),
        }
    }
}
