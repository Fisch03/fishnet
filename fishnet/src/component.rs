//! reusable components for building web pages (aka the main thing)

pub mod prelude;

mod build;
pub use build::{BuildableComponent, BuiltComponent, ComponentBuildResult};

mod render;
use render::ContentRenderer;

pub mod fake_macros;

use crate::css::StyleFragment;
use crate::js::ScriptType;
use crate::routes::ComponentRoute;

use axum::{
    body::Body, http::Request, response::IntoResponse, routing::method_routing::MethodRouter,
    Router,
};
use core::convert::Infallible;
use futures::future::BoxFuture;
use maud::Markup;
use std::{fmt::Debug, marker::PhantomData, ops::Deref};
use tower_service::Service;

#[derive(Debug, Clone)]
pub struct ComponentState<ST>
where
    ST: Clone + Send + Sync,
{
    api_route: ComponentRoute,
    state: ST,
}
impl<ST> ComponentState<ST>
where
    ST: Clone + Send + Sync,
{
    pub fn endpoint(&self) -> &str {
        self.api_route.as_str()
    }
}
impl<ST> Deref for ComponentState<ST>
where
    ST: Clone + Send + Sync,
{
    type Target = ST;
    fn deref(&self) -> &Self::Target {
        &self.state
    }
}

pub type ComponentRunner<ST> =
    Box<dyn FnOnce(ComponentState<ST>) -> BoxFuture<'static, ()> + Send + Sync + 'static>;

#[doc(hidden)]
pub struct NoRenderer;
#[doc(hidden)]
pub struct HasRenderer;
#[doc(hidden)]
pub struct NoState;
#[doc(hidden)]
pub struct FixedNoState;
#[doc(hidden)]
pub struct HasState;
pub struct Component<R, S, ST>
where
    ST: Clone + Send + Sync,
{
    name: String,
    id: String,
    is_dynamic: bool,

    renderer: Option<ContentRenderer<ST>>,

    state: ST,
    router: Option<Router>,

    runner: Option<ComponentRunner<ST>>,
    scripts: Vec<ScriptType>,
    style: Option<StyleFragment<'static>>,

    _renderer_state: PhantomData<R>,
    _state_state: PhantomData<S>,
}

// ---- constructor ----
impl Component<NoRenderer, NoState, ()> {
    pub fn new(name: &str, id: &str) -> Component<NoRenderer, NoState, ()> {
        Self {
            name: name.to_string(),
            id: id.to_string(),

            is_dynamic: false,

            state: (),
            router: None,

            renderer: None,
            runner: None,

            scripts: Vec::new(),
            style: None,

            _renderer_state: PhantomData,
            _state_state: PhantomData,
        }
    }
}

// ---- on all components ----
impl<R, S, ST> Component<R, S, ST>
where
    ST: Clone + Send + Sync,
{
    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn add_script(mut self, script: ScriptType) -> Self {
        self.scripts.push(script);
        self
    }

    pub fn style(mut self, style: StyleFragment<'static>) -> Self {
        self.style = Some(style);
        self
    }

    pub fn route(mut self, path: &str, route: MethodRouter) -> Component<R, S, ST> {
        let router = self.router.unwrap_or_else(|| Router::new());
        self.router = Some(router.route(path, route));
        self
    }

    pub fn nest_service<T>(mut self, path: &str, service: T) -> Self
    where
        T: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
        T::Response: IntoResponse,
        T::Future: Send + 'static,
    {
        let router = self.router.unwrap_or_else(|| Router::new());
        self.router = Some(router.nest_service(path, service));
        self
    }
}

// ---- adding a renderer ----
impl<S, ST> Component<NoRenderer, S, ST>
where
    ST: Clone + Send + Sync,
    S: Send + Sync + 'static,
{
    pub fn render<C>(self, renderer: C) -> impl BuildableComponent
    where
        ST: Clone + Send + Sync + 'static,
        C: Fn(ComponentState<ST>) -> BoxFuture<'static, Markup> + Send + Sync + 'static,
    {
        Component::<HasRenderer, S, ST> {
            name: self.name,
            id: self.id,

            is_dynamic: false,

            state: self.state,
            router: self.router,

            renderer: Some(Box::new(renderer)),
            runner: self.runner,

            scripts: self.scripts,
            style: self.style,

            _renderer_state: PhantomData,
            _state_state: PhantomData,
        }
    }

    pub fn render_dynamic<C>(self, renderer: C) -> impl BuildableComponent
    where
        ST: Clone + Send + Sync + 'static,
        C: Fn(ComponentState<ST>) -> BoxFuture<'static, Markup> + Send + Sync + 'static,
    {
        Component::<HasRenderer, S, ST> {
            name: self.name,
            id: self.id,
            is_dynamic: true,
            state: self.state,
            router: self.router,
            renderer: Some(Box::new(renderer)),
            runner: self.runner,
            scripts: self.scripts,
            style: self.style,
            _renderer_state: PhantomData,
            _state_state: PhantomData,
        }
    }
}

// ---- adding a runner ----
impl<R> Component<R, FixedNoState, ()> {
    pub fn with_runner<F>(mut self, runner: ComponentRunner<()>) -> Self {
        self.runner = Some(Box::new(runner));
        self
    }
}
impl<R, ST> Component<R, HasState, ST>
where
    ST: Clone + Send + Sync + 'static,
{
    pub fn with_runner<F>(mut self, runner: F) -> Self
    where
        F: FnOnce(ComponentState<ST>) -> BoxFuture<'static, ()> + Send + Sync + 'static,
    {
        self.runner = Some(Box::new(runner));
        self
    }
}

// ---- setting state ----
impl<R> Component<R, NoState, ()> {
    /// Add state to the component.
    /// Since the state will be passed around a lot, it should be cheap to clone. This usually
    /// means wrapping it in an Arc or similar.
    pub fn with_state<ST>(self, state: ST) -> Component<NoRenderer, HasState, ST>
    where
        ST: Clone + Send + Sync + 'static,
    {
        Component::<NoRenderer, HasState, ST> {
            name: self.name,
            id: self.id,

            is_dynamic: self.is_dynamic,

            state,
            router: self.router,

            renderer: None, // this is fine because there is no renderer on the component yet
            runner: None,   // runners can also only be added after with_state

            scripts: self.scripts,
            style: self.style,

            _renderer_state: PhantomData,
            _state_state: PhantomData,
        }
    }
}
