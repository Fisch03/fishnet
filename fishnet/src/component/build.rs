use async_trait::async_trait;
use axum::{Extension, Router};
use futures::future::{BoxFuture, FutureExt};
use maud::{html, Markup};
use std::any::TypeId;
use std::sync::Arc;
use tracing::{debug, instrument, trace};

use super::{
    render::{ComponentStyle, ContentType, StatefulContentRenderer},
    Component, ComponentRoute, ComponentState, HasRenderer,
};
use crate::js::ScriptType;
use crate::page::render_context;

#[derive(Debug, Clone)]
pub struct BuiltComponent {
    type_id: TypeId,

    #[allow(dead_code)]
    name: String,
    #[allow(dead_code)]
    id: String,

    content: Arc<ContentType>,
}

pub struct ComponentBuildResult {
    pub built_component: BuiltComponent,

    pub runner: Option<BoxFuture<'static, ()>>,
    pub router: Option<(ComponentRoute, Router)>,
}

pub struct ComponentGlobals {
    pub scripts: Vec<ScriptType>,
    pub style: Option<ComponentStyle>,
}

impl BuiltComponent {
    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn id(&self) -> &str {
        &self.id
    }

    pub fn type_id(&self) -> TypeId {
        self.type_id
    }

    pub async fn render(&self) -> Markup {
        html! {
            div class=(self.name) { (self.content.render().await) }
        }
    }

    pub fn is_dynamic(&self) -> bool {
        match self.content.as_ref() {
            ContentType::Static(_) => false,
            _ => true,
        }
    }
}

#[async_trait]
pub trait BuildableComponent {
    fn name(&self) -> &str;
    fn id(&self) -> &str;

    async fn build(self: Self, base_route: &str) -> ComponentBuildResult;
}

#[async_trait]
impl<ST, S> BuildableComponent for Component<HasRenderer, S, ST>
where
    ST: Clone + Send + Sync + 'static,
    S: Send + Sync + 'static,
{
    fn name(&self) -> &str {
        &self.name
    }
    fn id(&self) -> &str {
        &self.id
    }

    #[instrument(name = "build_component", skip_all, fields(name = %self.name))]
    async fn build(self, base_route: &str) -> ComponentBuildResult {
        trace!("building state");
        let api_route = ComponentRoute::new(base_route, &self.name, &self.id);
        let state = ComponentState {
            api_route: api_route.clone(),
            state: self.state,
        };

        let router = self.router.map(|r| r.layer(Extension(state.clone())));

        let renderer = self.renderer.unwrap();

        let runner = self.runner.map(|runner| {
            let runner = (runner)(state.clone());
            runner.boxed()
        });

        let content;

        if !self.is_dynamic {
            trace!("pre-rendering static component");
            render_context::enter_temporary_render().await;
            let render = renderer(state.clone()).await;
            if !render_context::exit_temporary_render().await {
                debug!("detected dynamic child, making self dynamic");
                content = ContentType::Dynamic(StatefulContentRenderer::new(renderer, state));
            } else {
                content = ContentType::Static(render);
            }
        } else {
            content = ContentType::Dynamic(StatefulContentRenderer::new(renderer, state));
        }

        trace!("rendering component style");
        let style = self.style.map(|style| style.render(&self.name));

        render_context::global_store()
            .add(self.type_id, || ComponentGlobals {
                scripts: self.scripts,
                style,
            })
            .await;

        debug!("built component");
        ComponentBuildResult {
            built_component: BuiltComponent {
                type_id: self.type_id,
                name: self.name,
                id: self.id,
                content: Arc::new(content),
            },
            runner,
            router: router.map(|r| (api_route, r)),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::component::prelude::*;
    use crate::component::render::ContentType;

    #[tokio::test]
    async fn test_build_minimal() {
        let result = component!(TestComponent)
            .render(|_| {
                async {
                    html! { "test" }
                }
                .boxed()
            })
            .build("/")
            .await;

        assert_eq!(result.built_component.name(), "TestComponent");
        assert!(matches!(
            result.built_component.content.as_ref(),
            ContentType::Static(_)
        ));

        assert!(result.runner.is_none());
        assert!(result.router.is_none());
    }

    #[tokio::test]
    async fn test_build_minimal_dynamic() {
        let result = component!(TestComponent)
            .render_dynamic(|_| {
                async {
                    html! { "test" }
                }
                .boxed()
            })
            .build("/")
            .await;

        assert_eq!(result.built_component.name(), "TestComponent");
        assert!(matches!(
            result.built_component.content.as_ref(),
            ContentType::Dynamic(_)
        ));

        assert!(result.runner.is_none());
        assert!(result.router.is_none());
    }

    #[tokio::test]
    async fn test_build() {
        use axum::routing::get;

        let result = component!(TestComponent)
            .add_script(ScriptType::External("test.js".to_string()))
            .route("/route", get(|| async { "test" }))
            .render(|_| {
                async {
                    html! { "test" }
                }
                .boxed()
            })
            .build("/")
            .await;

        assert_eq!(result.built_component.name(), "TestComponent");
        assert!(matches!(
            result.built_component.content.as_ref(),
            ContentType::Static(_)
        ));
        assert!(result.router.is_some());

        /* it would be nice to verify the global store here
        assert_eq!(result.scripts.len(), 1);
        assert_eq!(
            result.scripts[0],
            ScriptType::External("test.js".to_string())
        );
        */
    }

    /*
    #[test]
    fn test_nested_dynamic() {
        todo!();
    }
    */
}
