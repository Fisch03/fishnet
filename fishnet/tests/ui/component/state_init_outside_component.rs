use fishnet::component::prelude::*;

struct MyState {}

fn main() {
    state_init!(Arc<MyState>);
}
