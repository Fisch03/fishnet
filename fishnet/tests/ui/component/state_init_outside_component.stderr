error: the state_init macro can only be used from inside a component! consider annotating your function with '#[component]' or '#[dyn_component]'
 --> tests/ui/component/state_init_outside_component.rs:6:5
  |
6 |     state_init!(Arc<MyState>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `state_init` (in Nightly builds, run with -Z macro-backtrace for more info)
